fix(ivy): incorrect ChangeDetectorRef injected into pipes used in component inputs

crisbeto opened the pull request:
When injecting a `ChangeDetectorRef` into a pipe, the expected result is that the ref will be tied to the component in which the pipe is being used. This works for most cases, however when a pipe is used inside a property binding of a component (see test case as an example), the current `TNode` is pointing to component's host so we end up injecting the inner component's view. These changes fix the issue by only looking up the component view of the `TNode` if the `TNode` is a parent.

This PR resolves FW-1419.

alfaproject commented on packages/core/src/render3/view_engine_compatibility.ts:
```diff
@@ -363,20 +363,21 @@ export function createContainerRef(
  * Creates a ViewRef and stores it on the injector as ChangeDetectorRef (public alias).
  *
  * @param hostTNode The node that is requesting a ChangeDetectorRef
+ * @param hostTNodeIsParent Wheether the `hostTNode` is a parent node
```
```suggestion
 * @param hostTNodeIsParent Whether the `hostTNode` is a parent node
```

LayZeeDK commented:
Stumbled upon by @BioPhoton. Investigated by @JoostK.

Thanks for fixing!

kara commented on packages/core/src/render3/view_engine_compatibility.ts:
```diff
@@ -363,20 +363,21 @@ export function createContainerRef(
-  if (isComponent(hostTNode)) {
+function createViewRef(hostTNode: TNode, hostTNodeIsParent: boolean, hostView: LView, context: any):
+    ViewEngine_ChangeDetectorRef {
+  if (isComponent(hostTNode) && hostTNodeIsParent) {
```
Hmm this doesn't look quite right to me. The component `TNode` will be a parent if it has any content children, so I think if you added a div to your test, it wouldn't pass (can we add a separate test for that?). This works for the no-children case because you get an `element()` instruction that contains `elementEnd`.

It seems like we need some way to check that the requesting instruction is a directive and not a pipe before getting the component view 🤔 Perhaps we can generate a different instruction/arg for pipes instead of `injectChangeDetectorRef`?

crisbeto replied:
You're right, adding a child to the element breaks it. Regarding alternate solutions, I see the following alternatives:

1. As you mentioned, generating slightly different code for pipes. We'd probably need something called `pipeInject` to be generated instead of `directiveInject`. This should be the cleaner option, although it feels like a lot just for one injection token.
2. Have an extra parameter on `directiveInject` that marks it as a pipe. This feels weird, because we'd be generating the extra code even if `ChangeDetectorRef` isn't being injected.
3. Have some kind of flag in the global state similar to `isParent` that says whether we're currently creating a pipe (e.g. `getIsPipe` and `setIsPipe`). I tried it out and it works, but I don't know what the sentiment is around adding more things to the state.

What do you think?

kara replied:
Great analysis 👍 Agree that all options feel a bit heavy for just one case. I think #2 is out because it would generate so much extra code. 

Another strategy for #1 might be to special case `ChangeDetectorRef` in the compiler as another `R3ResolvedDependencyType` (like we do for attributes) and add an arg for `isPipe` to `compileFactoryFunction`. That way, we could generate `injectPipeChangeDetectorRef` instead of `directiveInject(ChangeDetectorRef)` if it's a pipe factory and CDR token specifically. What do you think about trying that? I'd prefer not to add a global unless we have to. But if it ends up getting more complicated, we can fall back to that?

crisbeto replied:
The alternate strategy for 1 makes sense. I'll try it out and see how far I get.

crisbeto committed:
fix(ivy): incorrect ChangeDetectorRef injected into pipes used in component inputs

When injecting a `ChangeDetectorRef` into a pipe, the expected result is that the ref will be tied to the component in which the pipe is being used. This works for most cases, however when a pipe is used inside a property binding of a component (see test case as an example), the current `TNode` is pointing to component's host so we end up injecting the inner component's view. These changes fix the issue by only looking up the component view of the `TNode` if the `TNode` is a parent.

This PR resolves FW-1419.

crisbeto commented:
@kara I've reworked it based on our discussion. Can you take another look?

kara reviewed (approved):
LGTM

kara commented:
[VE presubmit](https://test.corp.google.com/ui#id=OCL:258657523:BASE:258657546:1563403585940:20d55b76)

[Ivy presubmit](http://test/OCL:258657523:BASE:258657689:1563403641847:244d72e8)

kara commented on packages/compiler-cli/src/ngtsc/annotations/src/util.ts:
```diff
@@ -49,6 +49,14 @@ export function getConstructorDependencies(
+    if (typeValueReference && !typeValueReference.local &&
+        typeValueReference.name === 'ChangeDetectorRef' &&
+        typeValueReference.moduleName === '@angular/core') {
+      resolved = R3ResolvedDependencyType.ChangeDetectorRef;
```
G3 presubmit is throwing on `typeValueReference.name`. 

```ts
packages/compiler-cli/src/ngtsc/annotations/src/util.ts:55:28 - error TS2339: Property 'name' does not exist on type 'TypeValueReference'.
  Property 'name' does not exist on type '{ local: true; expression: Expression; defaultImportStatement: ImportDeclaration; }'.
```

Can we try something closer to the original resolver code for `ChangeDetectorRef` (e.g. https://github.com/angular/angular/pull/26064/files#diff-1a563f12050fa3c8eeab248f693dfa48L68)?

crisbeto commented:
Reworked based on the latest comment @kara. Can you take another look? All the changes are in the last fixup commit.

kara commented on packages/core/src/render3/instructions/di.ts:
```diff
@@ -59,3 +61,17 @@ export function ɵɵdirectiveInject<T>(
+ *
+ * @codeGenApi
+ */
+export function ɵɵinjectPipeChangeDetectorRef(flags = InjectFlags.Default): ChangeDetectorRef|null {
```
Type issue seems resolved. Now getting a cyclical dependency error in G3:

```ts
packages/core/src/render3/di.ts -> packages/core/public_api.ts -> packages/core/src/core.ts -> packages/core/src/metadata.ts -> packages/core/src/di.ts -> packages/core/src/di/index.ts -> packages/core/src/di/injectable.ts -> packages/core/src/di/jit/injectable.ts -> packages/core/src/di/jit/util.ts -> packages/core/src/change_detection/change_detector_ref.ts -> packages/core/src/render3/view_engine_compatibility.ts -> packages/core/src/render3/di.ts
```

Seems like we should move this function into `view_engine_compatibility` to resolve it.

crisbeto commented:
Reworked to avoid the circular import @kara.

kara commented:
Still getting a cyclic dep error:

```

packages/core/src/render3/di.ts -> packages/core/public_api.ts -> packages/core/src/core.ts -> 
packages/core/src/metadata.ts -> packages/core/src/di.ts -> packages/core/src/di/index.ts -> 
packages/core/src/di/injectable.ts -> packages/core/src/di/jit/injectable.ts -> 
packages/core/src/di/jit/util.ts -> packages/core/src/change_detection/change_detector_ref.ts -> 
packages/core/src/render3/view_engine_compatibility.ts -> packages/core/src/render3/di.ts
```

crisbeto commented:
Updated again @kara. For future reference, is there a way I can check this myself so we don't have to go back and forth as much?

AndrewKushnir commented on packages/core/src/render3/view_engine_compatibility.ts:
```diff
@@ -360,24 +360,24 @@ export function createContainerRef(
-    hostTNode: TNode, hostView: LView, context: any): ViewEngine_ChangeDetectorRef {
-  if (isComponent(hostTNode)) {
+function createViewRef(
+    hostTNode: TNode, hostView: LView, isPipe: boolean): ViewEngine_ChangeDetectorRef {
```
@crisbeto it looks like inline docs and function args are not aligned.

crisbeto replied:
Ah thanks, I forgot to update it after refactoring.

AndrewKushnir replied:
Thanks @crisbeto 👍

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/annotations/src/util.ts:
```diff
@@ -79,6 +80,11 @@ export function getConstructorDependencies(
       }
     });
+
+    if (token instanceof ExternalExpr && token.value.name === 'ChangeDetectorRef' &&
```
@crisbeto does it make sense to use `isAngularDecorator` function (defined below in `util.ts`) here?

crisbeto replied:
This isn't looking at a decorator, but at the type of the parameter.

crisbeto committed:
fixup! fix(ivy): incorrect ChangeDetectorRef injected into pipes used in component inputs

kara commented:
[new VE presubmit](https://tap.corp.google.com/ui#id=OCL:258852199:BASE:259465008:1563855008628:54d66b68)

[new Ivy presubmit](https://test.corp.google.com/ui#id=OCL:258852199:BASE:259465065:1563855056672:ff89ed24)

kara reviewed (approved):
LGTM

kara commented:
merge-assistance: global approval

LayZeeDK commented:
Was it merged?
